src/esp_loader.ts

```diff
@@ -768,6 +771,281 @@ export class ESPLoader extends EventTarget {
+
+    if (offset != this.getBootloaderOffset()) {
+      return image; // not flashing bootloader offset, so don't modify this image
+    }
+
+    // easy check if this is an image: does it start with a magic byte?
+    if (headerMagic != ESP_IMAGE_MAGIC) {
+      this.logger.log(
+        "Warning: Image file at %s doesn't look like an image file, so not changing any flash settings.",
+        toHex(offset, 4)
+      );
+      return image;
+    }
+
+    // make sure this really is an image, and not just data that
+    // starts with esp.ESP_IMAGE_MAGIC (mostly a problem for encrypted
+    // images that happen to start with a magic byte
+
+    // TODO Implement this test from esptool.py
+    /*
+    try:
+        test_image = esp.BOOTLOADER_IMAGE(io.BytesIO(image))
+        test_image.verify()
+    except Exception:
+        print("Warning: Image file at 0x%x is not a valid %s image, so not changing any flash settings." % (address, esp.CHIP_NAME))
+        return image
+    */
+
+    this.logger.log("Image being flashed is a bootloader");
+
+    let flashMode = FLASH_MODES["dio"]; // For now we always select dio, a common value supported by many flash chips and ESP boards
+    let flashFreq = FLASH_FREQUENCIES["40m"]; // For now we always select 40m, a common value supported by many flash chips and ESP boards
+    let flashSize =
+      this.getFlashSizes()[this.flashSize ? this.flashSize : "4MB"]; // If size was autodetected we use it otherwise we default to 4MB
+    let flashParams = pack("BB", flashMode, flashSize + flashFreq);
+    let imageFlashParams = new Uint8Array(image, 2, 2);
+
+    if (
+      flashParams[0] != imageFlashParams[0] ||
+      flashParams[1] != imageFlashParams[1]
+    ) {
+      imageFlashParams[0] = flashParams[0];
+      imageFlashParams[1] = flashParams[1];
+
+      this.logger.log(
+        `Patching Flash parameters header bytes to ${toHex(
+          flashParams[0],
+          2
+        )} ${toHex(flashParams[1], 2)}`
+      );
+    } else {
+      this.logger.log("Flash parameters header did not need patching.");
+    }
+    return image;
+  }
+
+  getFlashSizes() {
+    if (
+      this.chipFamily == CHIP_FAMILY_ESP32 ||
+      this._parent?.chipFamily == CHIP_FAMILY_ESP32
+    ) {
+      return ESP32_FLASH_SIZES;
+    }
+    return FLASH_SIZES;
+  }
+
+  async flashId() {
+    let SPIFLASH_RDID = 0x9f;
+    let result = await this.runSpiFlashCommand(SPIFLASH_RDID, [], 24);
+    return result;
+  }
+
+  getChipFamily() {
+    return this._parent ? this._parent.chipFamily : this.chipFamily;
+  }
+
+  async writeRegister(
+    address: number,
+    value: number,
+    mask = 0xffffffff,
+    delayUs = 0,
+    delayAfterUs = 0
+  ) {
+    let buffer = pack("<IIII", address, value, mask, delayUs);
+    if (delayAfterUs > 0) {
+      // add a dummy write to a date register as an excuse to have a delay
+      buffer.concat(
+        pack(
+          "<IIII",
+          getUartDateRegAddress(this.getChipFamily()),
```

Comment:
Since chip family is made up values, shouldn't we just pick a fixed value to "waste" a write on?

Reply:
The values of the "enum" are made up, but you still need to be able to tell them apart, in the code above getUartDateRegAddress returns the address for the UART DATE registry, it happens to be the same for the three chip families that we currently support, but it was different for others like the ESP32S3 or the ESP32C6

Reply:
Ah, I didn't realize `getUartDateRegAddress` returns a value based on the chip family. ok!